File: Recursion/Arrays/5.Palindrome_String.cpp
// Given a string, the task is to write a recursive function to check if the given string is palindrome or not.

// Examples:

// Input : string = "malayalam"
// Output : Yes
// Reverse of malayalam is also
// malayalam.

// Input : string = "max"
// Output : No
// Reverse of max is not max.

#include <iostream>
#include <string>
using namespace std;

bool isPalindrome(const string& str, int left, int right) {
    if (left >= right) return true; // Base Case: If pointers cross, it's a palindrome
    if (tolower(str[left]) != tolower(str[right]) ) return false; // Mismatch found, not a palindrome
    return isPalindrome(str, left + 1, right - 1); // Recursive call
}

int main() {
    string str;
    cin >> str; // Input string

    if (isPalindrome(str, 0, str.length() - 1)) {
        cout << "Palindrome" << endl;
    } else {
        cout << "Not a Palindrome" << endl;
    }

    return 0;
}

// Sample I/P & O/P:
// racecar
// Palindrome

// hello
// Not a Palindrome

File: Python/DSA/Arrays/Recursion/5.Palindrome_string.py
def is_palindrome(s: str, left: int, right: int) -> bool:
    """
    Recursive check if a string is a palindrome.

    Time Complexity (TC): O(n)
        - At each recursive step, we compare two characters and move inward.
        - At most n/2 recursive calls → O(n)

    Space Complexity (SC): O(n)
        - Due to recursion stack depth (up to n/2 calls)
    """
    # Base case: if pointers cross or meet, it's a palindrome
    if left >= right:
        return True
    # If characters mismatch, it's not a palindrome
    if s[left] != s[right]:
        return False
    # Move towards the center
    return is_palindrome(s, left + 1, right - 1)

if __name__ == "__main__":
   s = "malayalam"
   if is_palindrome(s, 0, len(s) - 1):
      print("Yes")
   else:
      print("No")
